File: 1 - BackEnd/Vanilla/PHP/estructuras_de_control/funciones/1 - funciones.php
<?php

/* ##########==================================########## */
/* ######===--- Declaracion e inicializacion ---===###### */
/* ##########==================================########## */

// Las funciones declaradas con la palabra reservada (function) se guardan en (pila).
// Las anonimas no se guardan en pila.
function funcion(){
	// ...
}

// Funcion (anonima).
// Esta funcion no se guarda en pila, sino por (declaracion comun).
$function = function(){
	// ...
};

// ------------------------- //
// ------ Llamamiento ------ //
// ------------------------- //

// Las funciones pueden mandar a llamar otras dentro de si mismas.
function funcion(){
	funcionUno();
	funcionDos();
}

funcion();

// ----------------------- //
// ------ Anidacion ------ //
// ----------------------- //

// Pueden haber funciones dentro de funciones y llamarse entre si.
function funcion(){
	$anidadaUno = function(){
		// ...
	};
	function anidadaDos(){
		// ...
	}

	anidadaUno();
	anidadaDos();
}

funcion();

/* ##########================########## */
/* ######===--- Parametros ---===###### */
/* ##########================########## */

// Aqui podemos tener tipado.
function decir_nombre($nombre, int $edad){
	// Recibimos los parametros y los imprimimos.
	printf("%s tiene una edad de: (%d).", $nombre, $edad);
}

decir_nombre("Brandon", 21);

// Funcion anonima.
$my_function = function(int $numero):void{
	for($i = 0; $i < $numero; $i++){
		printf("<div>Numero: %d hasta %d</div>", $i, $numero);
	}
};

$my_function(30);

// ------------------------------------- //
// ------ Valores predeterminados ------ //
// ------------------------------------- //

// Recibimos valores predefinidos.
$informacion = function(string $nombre = "Ninguno", int $edad = 21){
	echo("El usuario {$nombre} tiene la edad de: ".$edad);
};

// En este caso el segundo argumento ya lo toma por defecto como (21).
$informacion("Brandon");

// ----------------------------- //
// ------ Tipado opcional ------ //
// ----------------------------- //

// (?), indica que se recibe una propiedad de tipo (string o null).
function funcion(?string $name){
	// ...
}

// (?), indica que se puede retornar un dato de tipo (string o null).
function get_name():?string{
	return null;
}

// -------------------- //
// ------ Arrays ------ //
// -------------------- //

// Un arreglo (opcional).
$sumar = function(array $array = array()):void{ 
	$suma = 0;
	
	foreach($array as $indice => $valor){
		$suma += $valor;
	}

	echo("La suma es: ".$suma);
};

// Podemos pasarle un array y este sumara cada uno de sus valores.
$sumar([5,5,5]);
$sumar([
	10, 20, 5, 2, 1, 10
]);

/**
 * Otros parametros.
 */
$mostrar_usuarios = function(array $arr_usuarios = [], string $mensaje = "No hay mas usuarios"){

	printf("<div><strong>%s</strong></div>", $mensaje);
    
    for($i = 0; $i < count($arr_usuarios); $i++){
    	echo("<div>{$arr_usuarios[$i]}</div>");
    }

};

$mostrar_usuarios([
	"Brandon", "Anthony", "Cosmic"
]);

$mostrar_usuarios(
	["Brandon", "Anthony", "Cosmic"], 
	"Son todos los usuarios por ahora."
);

// --------------------- //
// ------ Objetos ------ //
// --------------------- //

function obtener($usuario){
	// Aqui podemos recivir cualquier tipo de dato.
}

function objeto(object $objeto){
	// Especifica que debe recibir un objeto, (sin importar el tipo de objeto que sea).
}

function obtener_typado(Usuario $usuario){
	// Aqui se declara que el parametro debe ser una instancia de la clase (Usuario).
}

/* ##########======================########## */
/* ######===--- Retorno de datos ---===###### */
/* ##########======================########## */

function suma(int $numero_0 = 0, int $numero_1 = 0){
	return ($numero_0 + $numero_1);
}

// Podemos pasar algunos parametros ya definidos y trabajar con ellos.
$to_fixed = function(float $numero = 0.000, int $decimales = 0){
	$num_texto = ((string) $numero);
	$num_final = "";
	$decimales++;

	if(str_contains($num_texto, ".")){
		$posicion = strpos($num_texto, ".")+$decimales;
		$num_final = substr($num_texto, 0, $posicion);
	}else{
		return 0;
	}
	return ((float) $num_final);
};

$to_fixed(5.553252525425,3); // 5.553
$to_fixed(); // 0
$to_fixed(555.555); // 555

// -------------------- //
// ------ Arrays ------ //
// -------------------- //

$etiquetar_usuarios = function($arr_usuarios = []){
	$etiquetados = [];

	foreach($arr_usuarios as $indice => $key){
		array_push($etiquetados, "(Usuario => {$key})");
	}

	// Retornamos el array nuevo.
	return $etiquetados;
};

$etiquetar_usuarios(array(
	"nombre_0" => "Brandon", 
	"nombre_1" => "Anthony", 
	"nombre_2" => "Cosmic"
));

// ----------------------- //
// ------ Funciones ------ //
// ----------------------- //

function obtener(){
	return function(){
		echo "Hola mundo";
	};
}

// Guardar la funcion devuelta de forma anonima e invocarla.
$funcion = obtener(); $funcion();

// Invocarla directamente.
obtener()();

###### ------------------------------------------------------------------------------- ######

function obtener(callback){
	return callback();
}

// Se recibe la funcion anonima.
obtener(function(){
	// Se retorna otra funcion.
	return function(){
		echo "Hola mundo";
	};

// Se manda a llamar la funcion retornada, ().
})();

// --------------------- //
// ------ Objetos ------ //
// --------------------- //

// Puede devolver cualquier tipo de dato.
function sinEspecificar(){
	return [
		new Unknown(), 
		new Usuario()
	];
}

// Especifica que puede retornar cualquier objeto.
function retornar():object{
	return new Unknown();
}

// Retorna un tipado especifico.
function retornar_usuario():Usuario{
	return new Usuario();
}

/* ##########==============================########## */
/* ######===--- Referencias de variables ---===###### */
/* ##########==============================########## */

$cursos = 'PHP';

# Con el '&' hace referencia a la otra variable que se encuentra fuera.
function path(&$cursos){
	$cursos = 'Laravel';
	echo $cursos;
}

# Ambos terminan imprimiendo 'Laravel'.
path($cursos);
echo $cursos;

// -------------------------------- //
// ------ Varialbes globales ------ //
// -------------------------------- //

$variable_global = true;

function funcion(){
	// De esta manera puede acceder a las variables fuera de ella.
	global $variable_global;
}

/* ##########==================########## */
/* ######===--- Recursividad ---===###### */
/* ##########==================########## */

# Creamos una variable global para llevar el conteo.
$GLOBALS['numero'] = 0;


function recursividad($cadena = "", $numero = 0){
	// Cuando la condicion se cumple finaliza.
	if($GLOBALS['numero'] >= $numero){
		echo "<b>fin</b>";
		return false;
	}else{
		$GLOBALS['numero']++;
		print $GLOBALS['numero'];

		// Se llama a si misma.
		recursividad($cadena, $numero);
	}
}

recursividad("El numero es: ", 10);
/**
 * El numero es: 1
 * El numero es: 2
 * El numero es: 3
 * El numero es: 4
 * El numero es: 5
 * El numero es: 6
 * El numero es: 7
 * El numero es: 8
 * El numero es: 9
 * El numero es: 10
 * Bucle finalizado.
 */

/* ##########==============########## */
/* ######===--- Callback ---===###### */
/* ##########==============########## */

/**
 * NOTA: Las funciones pueden aceptar (funciones anonimas) por (callbacks).
 */

// Recibe un (callback) que retorna algo y ese algo se valida dentro.

function sumar(int $numero_0 = 0, int $numero_1 = 0):int{
	return ($numero_0 + $numero_1);
}

$llamar = function($callback):bool{
	if($callback() > 10) return true;
	else return false;
}

// Llmamiento de la funcion y sus (callbacks).
$llamar(function(){
	return sumar(5, 5);
});

// -------------------------------------------- //
// ------ La palabra reservada (Closure) ------ //
// -------------------------------------------- //

# 'Closure' hace que forsozamente se pase una funcion anonima.
function greet(Closure $lang, $name){
	return $lang($name);
}

$es = function($name){
	return "Hola, $name";
};

$en = function(string $name):string{
	return "Hello, $name";
};

echo greet($es, "Lynda");

File: 1 - BackEnd/Vanilla/Python/estructuras-de-control/funciones/1 - funciones.py
# ##########==================================##########
# ######===--- Declaración e inicialización ---===######
# ##########==================================##########

# NOTA: (Placeholder) y (None): Indica que el codigo esta vacio, sino dara error.

# Las funciones se definen con la palabra reservada (def).
def funcion():
    # ...
    pass  # Placeholder para el contenido de la función

# Función anónima (lambda).
# Estas funciones son de una sola línea y no se almacenan en pila como las normales.
funcion_anomima = lambda: None  # Placeholder para una función anónima
suma = lambda x, y: x + y

# Llamando a la función anónima
suma(5, 3) # 8

# ------------------------- //
# ------ Llamamiento ------ //
# ------------------------- //

# Las funciones pueden llamar a otras dentro de sí mismas.
def funcion_principal():
    funcion_uno()
    funcion_dos()

funcion_principal()  # Llama a la función principal

# ----------------------- //
# ------ Anidación ------ //
# ----------------------- //

# Pueden haber funciones dentro de funciones y llamarse entre sí.
def funcion_anidada():
    def anidada_uno():
        print("Función Anidada Uno")

    def anidada_dos():
        print("Función Anidada Dos")

    anidada_uno()  # Llama a la función anidada uno
    anidada_dos()  # Llama a la función anidada dos

funcion_anidada()  # Llama a la función que contiene las anidadas

# ##########================##########
# ######===--- Parámetros ---===######
# ##########================##########

# Aquí podemos tener tipado (usando type hints).
#   --- (-> None), es el retorno.
def decir_nombre(nombre:str, edad:int)->None:
    # Recibimos los parámetros y los imprimimos.
    print(f"{nombre} tiene una edad de: {edad}.")

decir_nombre("Brandon", 21)

# Función anónima.
my_function = lambda numero: [
    print(f"Numero: {i} hasta {numero}") 
    for i in range(numero)
]

my_function(30)

# ------------------------------------- #
# ------ Valores predeterminados ------ #
# ------------------------------------- #

# Recibimos valores predefinidos.
def informacion(nombre:str = "Ninguno", edad:int = 21)->None:
    print(f"El usuario {nombre} tiene la edad de: {edad}.")

# En este caso el segundo argumento ya lo toma por defecto como (21).
informacion("Brandon")

# ----------------------------- #
# ------ Tipado opcional ------ #
# ----------------------------- #

# Importar para hacer el opcional similar a Typescript (?).
from typing import Optional

def funcion(name:Optional[str])->None:
    # ...
    pass

def get_name()->Optional[str]:
    return None

# -------------------- #
# ------ Listas ------ #
# -------------------- #

# Una lista (opcional).
def sumar(lista:list = [])->None: 
    print(f"La suma es: {sum(lista)}")

# Podemos pasarle una lista y esta sumará cada uno de sus valores.
sumar([5, 5, 5])
sumar([10, 20, 5, 2, 1, 10])

# Otros parámetros.
def mostrar_usuarios(arr_usuarios:list = [], mensaje:str = "Sin usuarios")->None:
    for usuario in arr_usuarios:
        print(f"usuario\n")

mostrar_usuarios(["Brandon", "Anthony", "Cosmic"])

mostrar_usuarios(
    ["Brandon", "Anthony", "Cosmic"], 
    "Son todos los usuarios por ahora."
)

# --------------------- #
# ------ Objetos ------ #
# --------------------- #

# Clase para nuestro tipado.
class Usuario:
    pass

'''
    Funciones
'''

def obtener(usuario)->None:
    # Aquí podemos recibir cualquier tipo de dato.
    pass

def objeto(objeto:object)->None:
    # Especifica que debe recibir un objeto (sin importar el tipo de objeto que sea).
    pass

def obtener_typado(usuario:Usuario)->None:
    # Aquí se declara que el parámetro debe ser una instancia de la clase Usuario.
    pass

# ##########======================##########
# ######===--- Retorno de datos ---===######
# ##########======================##########

# Función que retorna la suma de dos números.
def suma(numero_0:int = 0, numero_1:int = 0) -> int:
    return numero_0 + numero_1

# Ejemplo de uso
suma(5, 3)  # 8

# Función para redondear a un número de decimales específicos.
def to_fixed(numero:float = 0.0, decimales:int = 0)->float:
    num_texto = str(numero)
    num_final = ""

    # Aumentar decimales para incluir el redondeo.
    decimales += 1

    if "." in num_texto:
        posicion = num_texto.index(".") + decimales
        num_final = num_texto[:posicion]
    else:
        return 0.0

    return float(num_final)

to_fixed(5.553252525425, 3) # 5.553
to_fixed() # 0.0
to_fixed(555.555) # 555.0

# -------------------- #
# ------ Listas ------ #
# -------------------- #

# Función para etiquetar usuarios en una lista.
def etiquetar_usuarios(arr_usuarios:list = [])->list:
    etiquetados = []

    for key in arr_usuarios:
        etiquetados.append(f"(Usuario => {key})")

    # Retornamos la lista nueva.
    return etiquetados

etiquetar_usuarios(["Brandon", "Anthony", "Cosmic"])

# ----------------------- #
# ------ Funciones ------ #
# ----------------------- #

# Función que devuelve otra función.
def obtener():
    return lambda: print("Hola mundo")

# Guardar la función devuelta de forma anónima e invocarla.
funcion = obtener()
funcion()  # "Hola mundo"

# Invocarla directamente.
obtener()()  # "Hola mundo"

# --------------------- #
# ------ Objetos ------ #
# --------------------- #

class Unknown:
    pass

class Usuario:
    pass

# Puede devolver cualquier tipo de dato.
def sin_especificar()->list:
    return [Unknown(), Usuario()]

# Especifica que puede retornar cualquier objeto.
def retornar() -> object:
    return Unknown()

# Retorna un tipado específico.
def retornar_usuario() -> Usuario:
    return Usuario()

# ##########==============================##########
# ######===--- Referencias de variables ---===######
# ##########==============================##########

# -------------------------------- //
# ------ Variables globales ------ //
# -------------------------------- //

variable_global = True

def funcion():
    global variable_global  # Declaramos que usaremos la variable global
    print(variable_global)  # Accedemos a la variable global

funcion()  # True

# ##########==================##########
# ######===--- Recursividad ---===######
# ##########==================##########

# Creamos una variable global para llevar el conteo.
numero = 0

def recursividad(cadena="", limite=0):
    global numero  # Usamos la variable global
    
    # Cuando la condición se cumple finaliza.
    if numero >= limite:
        print("Fin")
        return False
    else:
        numero += 1
        print(cadena + str(numero))

        # Se llama a sí misma.
        recursividad(cadena, limite)

recursividad("El número es: ", 10)
'''
    El número es: 1
    El número es: 2
    El número es: 3
    El número es: 4
    El número es: 5
    El número es: 6
    El número es: 7
    El número es: 8
    El número es: 9
    El número es: 10
    Fin
'''

# ##########==============##########
# ######===--- Callback ---===######
# ##########==============##########

"""
    NOTA: Las funciones pueden aceptar (funciones anónimas) como (callbacks).
"""

def sumar(numero_0=0, numero_1=0):
    return numero_0 + numero_1

def llamar(callback):
    # Mandamos a llamar la (funcion).
    if callback() > 10:
        return True
    else:
        return False

# Llamamiento de la función y le pasamos el (callback) por parametro.
resultado = llamar(lambda: sumar(5, 5))